### DISCUSSION TITLE
Documentation on VM communication

### DISCUSSION THREAD

**Discussion by davidchisnall** [containers/podman#18490] [Q&A]
I am using Podman on FreeBSD and it seems to work very well. I’d like to be able to run FreeBSD containers in VMs on macOS and Windows with Podman Desktop. As I understand it, Podman Desktop starts a Linux VM image to run Linux VMs and could do the same for FreeBSD. Are the requirements for the VM image documented somewhere (what does it need to run, how does it communicate with the host, and so on)?

**Comment by benoitf** [comment]
hello @davidchisnall  I will transfer the question to the podman repository as 'podman machines' are provided through this repository

**Reply by davidchisnall** [reply]
Thanks.

**Comment by afbjorklund** [comment]
I think you can run any OS, as long as it is CoreOS (with "ignition" and everything)?

There were some FreeBSD experiments done in the Lima project recently, maybe easier...

* https://github.com/lima-vm/lima/issues/1508

It needs a ZFS disk, for storage.

**Comment by davidchisnall** [comment]
To expand a bit:

I am currently able to create a FreeBSD VM manually, deploy podman in it, and then connect to it remotely and manage FreeBSD containers from my Mac.  I would like to automate this via the `podman machine init` command but I'm struggling somewhat to understand the layering and the separation of concerns between the generic code and the hypervisor-specific bits.  For example, it's not clear to me how the SSH keys are provisioned to the guest.  It looks as if some things come via an Ignition file, which looks to be a Fedora-specific thing (though it's JSON, so writing a tool that handles the subset that I care about would be fairly trivial), but this seems to come via the QEMU fw_cfg device in a well-known location on QEMU, but I am not certain how it works elsewhere.

As a minimum, I would like to understand what the requirements are from an image provided by `--image-path` to be able to work (ignoring host volume mounts for now).

**Comment by davidchisnall** [comment]
From a bit of spelunking in the code, I think that there are two phases:

 - Provisioning initial SSH keys, which seems to be done in a hypervisor-specific way and using a RedHat-specific format (Ignition).  FreeBSD currently lacks a fw_cfg device, but that would be very easy to add.  I am not sure I follow the Hyper-V code well enough to understand how it's provided there.  Since this is a JSON document and the fields used in it are fairly restricted, it would be easy for me to add a tool to parse this and provision the SSH authorised key in a FreeBSD VM.

 - Sharing volumes.  This is done via 9p-over-virtio.  It's not yet clear to me how the guest is expected to understand the mount points.

There doesn't seem to be much by way of layering between the guest abstraction and the hypervisor abstraction.  For example, the Hyper-V code injects a systemd unit to detect when boot has finished and provide a notification over the serial terminal, so is assuming the guest is running systemd (definitely not the case for a FreeBSD guest, not universally true for a Linux guest).  The hypervisor drivers also appear to hard-code the default locations of the sockets, which I would expect to be part of the guest OS configuration.  I presume that this is an artefact of the current implementation where there are only two supported (similar) guests and so they are very similar, rather than an explicit design choice, but since I haven't been able to find any documentation other than the code, it's not really clear.

**Reply by davidchisnall** [reply]
I have now implemented the guest portion of the QEMU fw_cfg interface in FreeBSD:

https://github.com/davidchisnall/qemufwcfg

This should make it possible to get the Ignition file that `podman machine init` uses to provision SSH keys.
